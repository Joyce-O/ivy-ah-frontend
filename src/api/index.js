import axios from 'axios';

const API_BASE_URL = 'https://ivy-ah-backend-staging.herokuapp.com/api/v1';

const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6Iml2eWFkbWluQGdtYWlsLmNvbSIsImlkIjoiMDU1MDk2YWUtMDcyYy00MjJkLTgwMTctZDM2ZTFiZWNmNTNjIiwicm9sZSI6ImFkbWluIiwiaWF0IjoxNTU3MTUzNzc2fQ.SeUSBXKom7eHkEXgak8uS-3xqVKSwR3DvrGyIcn7CSs';

export const client = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
    Authorization: token
  },
});
export const signup = params => client.post('/users/signup', params);
export const login = params => client.post('/users/login', params);
export const fetchBookmarks = () => client.get('/bookmarks');
export const addBookmark = article => client.post('/bookmarks', { article });
export const removeBookmark = article => client.delete('/bookmarks', {
  data: {
    article
  }
});
export const fetchArticle = id => client.get(`/articles/${id}`);
