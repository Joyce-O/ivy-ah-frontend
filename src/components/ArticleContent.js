import React from 'react';
import { convertFromRaw, EditorState } from 'draft-js';
import { Editor } from 'react-draft-wysiwyg';
import PropTypes from 'prop-types';

const ArticleContent = ({ body }) => {
  const contentState = convertFromRaw(JSON.parse(body));
  return (
    <div>
      <Editor
        editorClassName="main-editor"
        editorState={EditorState.createWithContent(contentState)}
        toolbarHidden
        readOnly
      />
    </div>

  );
};

ArticleContent.propTypes = {
  body: PropTypes.string.isRequired
};

export default ArticleContent;
