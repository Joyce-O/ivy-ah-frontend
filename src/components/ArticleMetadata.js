import React from 'react';
import PropTypes from 'prop-types';
import hypeIcon from '../assets/sass/components/assets/hype-icon.svg';

const ArticleMetadata = ({
  data: {
    title, user, readTime, isPremium, createdAt, ratings,
  }
}) => (
  <div className="article-metadata">
    <h1 className="article-title">{title}</h1>
    {user.firstname
      ? (
        <span className="author-name">
          {user.firstname}
          {' '}
          {user.lastname}
        </span>
      )
      : <span className="author-name">{user.username}</span>}
    <div className="time-date">
      <span className="read-time dot">
        {`${Math.floor(parseInt(readTime, 10) / 60) || '< 1'} min read`}
      </span>
      <span className="published-date dot">{new Date(createdAt).toDateString()}</span>
    </div>
    <div className="ui horizontal link list hype-premium">
      <span className="item">
        <img src={hypeIcon} alt="hype-icon" className="hype article-icon" />
        <span className="hype-text">
          {ratings >= 1000 ? (ratings / 1000).toFixed(1) : ratings}
          {' Hypes'}
        </span>
      </span>
      {isPremium
          && (
            <span className="item">
              <i className="star icon article-icon yellow" />
            </span>
          )}
    </div>
  </div>
);

ArticleMetadata.propTypes = {
  data: PropTypes.shape({
    title: PropTypes.string,
    user: PropTypes.shape({
      firstname: PropTypes.string,
      lastname: PropTypes.string,
      username: PropTypes.string
    }),
    readTime: PropTypes.string,
    isPremium: PropTypes.bool,
    createdAt: PropTypes.string,
    ratings: PropTypes.number
  })
};

ArticleMetadata.defaultProps = {
  data: {
    title: '',
    user: {},
    readTime: '',
    isPremium: false,
    ratings: 0,
    createdAt: ''
  }
};


export default ArticleMetadata;
