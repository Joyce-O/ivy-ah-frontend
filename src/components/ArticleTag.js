import React from 'react';
import PropTypes from 'prop-types';

const ArticleTag = ({ tagList }) => (
  <React.Fragment>
    {tagList.map(tag => (
      <span className="article-tag" key={tag}>
        {tag}
      </span>
    ))}
  </React.Fragment>
);

ArticleTag.propTypes = {
  tagList: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string])).isRequired
};

export default ArticleTag;
