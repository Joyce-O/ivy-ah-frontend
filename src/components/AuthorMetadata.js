import React from 'react';
import PropTypes from 'prop-types';

const AuthorMetadata = ({
  user: {
    username, bio, image, firstname
  }
}) => (
  <div className="author-metadata">
    {image
      ? (
        <img className="profile-pic" src={image} alt={firstname} />
      )
      : (
        <img className="profile-pic" src="http://www.gravatar.com/avatar/?d=identicon" alt={firstname} />
      )
    }
    <div className="author-bio">
      {firstname
        ? <span className="author-name">{firstname}</span>
        : <span className="author-name">{username}</span>
      }
      <br />
      {bio && <span className="bio">{bio}</span>}
      <br />
      <button type="button" className="ui button follow">Follow</button>
    </div>
  </div>
);

AuthorMetadata.propTypes = {
  user: PropTypes.shape({
    username: PropTypes.string,
    bio: PropTypes.string,
    image: PropTypes.string,
    firstname: PropTypes.string,
  }).isRequired
};

export default AuthorMetadata;
