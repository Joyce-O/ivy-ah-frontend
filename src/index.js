import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import 'semantic-ui-css/semantic.min.css';
import 'overlayscrollbars/css/OverlayScrollbars.min.css';
import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import App from './containers/App';
import Store from './store';
import './assets/sass/index.scss';

render(<Provider store={Store}><App /></Provider>, document.getElementById('root'));
