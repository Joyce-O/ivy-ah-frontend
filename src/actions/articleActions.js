import * as api from '../api';
import * as actions from './actionTypes';

export const fetchArticleStarted = () => ({
  type: actions.FETCH_ARTICLE_STARTED
});

export const fetchArticleSucceeded = article => ({
  type: actions.FETCH_ARTICLE_SUCCEEDED,
  payload: { article }
});

export const fetchArticleFailed = error => ({
  type: actions.FETCH_ARTICLE_FAILED,
  payload: { error }
});

export const fetchArticle = id => (dispatch) => {
  dispatch(fetchArticleStarted());
  return api.fetchArticle(id)
    .then((response) => {
      dispatch(fetchArticleSucceeded(response.data));
    })
    .catch(error => dispatch(fetchArticleFailed(error)));
};
