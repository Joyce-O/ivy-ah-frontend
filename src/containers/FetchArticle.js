import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchArticle } from '../actions/articleActions';
import ArticlePage from '../components/ArticlePage';

export const Article = ({
  // eslint-disable-next-line no-shadow
  article, match, fetchArticle, loading, error
}) => {
  useEffect(() => {
    fetchArticle(match.params.id);
  }, []);

  return (
    <div className="article-view">
      <ArticlePage
        article={article}
        loading={loading}
        error={error}
      />
    </div>
  );
};

Article.propTypes = {
  article: PropTypes.shape({}),
  fetchArticle: PropTypes.func.isRequired,
  match: PropTypes.shape({}),
  loading: PropTypes.bool,
  error: PropTypes.bool
};

Article.defaultProps = {
  article: {},
  match: {},
  loading: true,
  error: false
};

const mapStateToProps = (state) => {
  const { article, loading, error } = state.article;
  return {
    article, loading, error
  };
};

const mapDispatchToProps = {
  fetchArticle
};

export default connect(mapStateToProps, mapDispatchToProps)(Article);
