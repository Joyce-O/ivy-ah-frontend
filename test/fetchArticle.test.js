import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createStore } from 'redux';
import * as api from '../src/api';
import { Article } from '../src/containers/FetchArticle';
import ArticleContent from '../src/components/ArticleContent';
import ArticleMetadata from '../src/components/ArticleMetadata';
import ArticleTag from '../src/components/ArticleTag';
import singleArticleReducer from '../src/reducers/singleArticleReducer';
import * as articleActions from '../src/actions/articleActions';
import AuthorMetadata from '../src/components/AuthorMetadata';


const mockStore = configureStore([thunk]);

describe('Test Article actions', () => {
  it('calls request and success actions if the fetch response was successful', () => {
    api.client.get = jest.fn().mockReturnValue(Promise.resolve({
      data: {
        body: JSON.stringify({}),
        tagList: ['Movies', 'Sport', 'Business'],
        user: {
          firstname: 'Jude',
          lastname: 'Afam',
          username: 'bobjayafam'
        }
      }
    }));

    const expectedActions = [
      'FETCH_ARTICLE_STARTED',
      'FETCH_ARTICLE_SUCCEEDED'
    ];
    const store = mockStore({});
    return store.dispatch(articleActions.fetchArticle('2aec1bef'))
      .then(() => {
        const actualActions = store.getActions().map(action => action.type);
        expect(actualActions).toEqual(expectedActions);
      });
  });

  it('calls request and success actions if the fetch response was successful', () => {
    api.client.get = jest.fn().mockReturnValue(Promise.reject(new Error('An error occured')));
    const expectedActions = [
      'FETCH_ARTICLE_STARTED',
      'FETCH_ARTICLE_FAILED'
    ];
    const store = mockStore({});
    return store.dispatch(articleActions.fetchArticle('2aec1bef'))
      .then(() => {
        const actualActions = store.getActions().map(action => action.type);
        expect(actualActions).toEqual(expectedActions);
      });
  });
});

describe('Test single article reducer', () => {
  const initialArticleState = {
    article: {},
    error: false,
    loading: false
  };

  it('should return loading when passed FETCH_ARTICLE_STARTED', () => {
    const action = articleActions.fetchArticleStarted();
    const newState = singleArticleReducer(initialArticleState, action);
    expect(newState).toEqual({
      article: {},
      error: false,
      loading: true
    });
  });
  it('should return an article when passed FETCH_ARTICLE_SUCCEEDED', () => {
    const article = {
      title: 'My first article'
    };
    const action = articleActions.fetchArticleSucceeded(article);
    const newState = singleArticleReducer(initialArticleState, action);
    expect(newState).toEqual({ loading: false, article, error: false });
  });

  it('should return an error when passed FETCH_ARTICLE_FAILED', () => {
    const error = true;
    const action = articleActions.fetchArticleFailed(error);
    const newState = singleArticleReducer(initialArticleState, action);
    expect(newState).toEqual({ article: {}, loading: false, error });
  });
});

describe('<Article />', () => {
  it('renders the component', () => {
    const props = {
      article: {
        data: {
          body: JSON.stringify({
            entityMap: {},
            blocks: [{
              key: '637gr',
              text: 'Initialized from content state.',
              type: 'unstyled',
              depth: 0,
              inlineStyleRanges: [],
              entityRanges: [],
              data: {}
            }]
          }),
          tagList: ['Movies', 'Sport', 'Business'],
          user: {
            firstname: 'Jude',
            lastname: 'Afam',
            username: 'bobjayafam'
          },
          title: 'My First Article',
          readTime: '76 secs',
          isPremium: true,
          createdAt: '2019-05-06T15:21:13.799Z',
          ratings: 100,
        },
      },
      match: {
        params: {}
      },
      fetchArticle: jest.fn()
    };
    const wrapper = mount(<Article {...props} />);
    expect(wrapper.find('h1').text()).toEqual('My First Article');
  });
});

describe('<ArticleContent />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const props = {
        body: JSON.stringify({
          entityMap: {},
          blocks: [{
            key: '637gr',
            text: 'Initialized from content state.',
            type: 'unstyled',
            depth: 0,
            inlineStyleRanges: [],
            entityRanges: [],
            data: {}
          }]
        }),
      };

      const wrapper = shallow(<ArticleContent {...props} />);

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});

describe('<ArticleMetadata />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const props = {
        data: {
          title: 'My First Article',
          readTime: '76 secs',
          isPremium: true,
          createdAt: '2019-05-06T15:21:13.799Z',
          ratings: 100,
          user: {
            firstname: 'Jude',
            lastname: 'Afam',
            username: 'bobjayafam'
          }
        }
      };

      const wrapper = shallow(<ArticleMetadata {...props} />);
      expect(wrapper.find('h1').text()).toEqual('My First Article');
      expect(wrapper.find('.author-name').text()).toEqual('Jude Afam');
    });
  });
});

describe('<ArticleTag />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const props = {
        tagList: ['Movies', 'Sport', 'Business']
      };

      const wrapper = shallow(<ArticleTag {...props} />);

      expect(wrapper.find('.article-tag').at(0).text()).toEqual('Movies');
    });
  });
});

describe('<AuthorMetadata />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const props = {
        user: {
          firstname: 'Jude',
          lastname: 'Afam',
          username: 'bobjayafam',
          bio: 'A web developer'
        }
      };

      const wrapper = shallow(<AuthorMetadata {...props} />);
      expect(wrapper.find('.author-name').text()).toEqual('Jude');
    });
  });
});

describe('Test Store', () => {
  it('should handle fetching a single article', () => {
    const initialArticleState = {
      article: {},
      error: false,
      loading: false
    };
    const article = {
      title: 'My first article'
    };
    const store = createStore(singleArticleReducer, initialArticleState);
    const action = articleActions.fetchArticleSucceeded(article);
    store.dispatch(action);
    const fetchedArticle = store.getState().article;
    expect(fetchedArticle).toEqual(article);
  });
});
