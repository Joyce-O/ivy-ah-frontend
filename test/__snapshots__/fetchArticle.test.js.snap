// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleContent /> render() renders the component 1`] = `
<div>
  <t
    customDecorators={Array []}
    editorClassName="main-editor"
    editorState={
      EditorState {
        "_immutable": Immutable.Record {
          "allowUndo": true,
          "currentContent": Immutable.Record {
            "entityMap": Object {},
            "blockMap": Immutable.OrderedMap {
              "637gr": Immutable.Record {
                "key": "637gr",
                "type": "unstyled",
                "text": "Initialized from content state.",
                "characterList": Immutable.List [
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                ],
                "depth": 0,
                "data": Immutable.Map {},
              },
            },
            "selectionBefore": Immutable.Record {
              "anchorKey": "637gr",
              "anchorOffset": 0,
              "focusKey": "637gr",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
            "selectionAfter": Immutable.Record {
              "anchorKey": "637gr",
              "anchorOffset": 0,
              "focusKey": "637gr",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
          },
          "decorator": null,
          "directionMap": Immutable.OrderedMap {
            "637gr": "LTR",
          },
          "forceSelection": false,
          "inCompositionMode": false,
          "inlineStyleOverride": null,
          "lastChangeType": null,
          "nativelyRenderedContent": null,
          "redoStack": Immutable.Stack [],
          "selection": Immutable.Record {
            "anchorKey": "637gr",
            "anchorOffset": 0,
            "focusKey": "637gr",
            "focusOffset": 0,
            "isBackward": false,
            "hasFocus": false,
          },
          "treeMap": Immutable.OrderedMap {
            "637gr": Immutable.List [
              Immutable.Record {
                "start": 0,
                "end": 31,
                "decoratorKey": null,
                "leaves": Immutable.List [
                  Immutable.Record {
                    "start": 0,
                    "end": 31,
                  },
                ],
              },
            ],
          },
          "undoStack": Immutable.Stack [],
        },
      }
    }
    localization={
      Object {
        "locale": "en",
        "translations": Object {},
      }
    }
    readOnly={true}
    stripPastedStyles={false}
    toolbarHidden={true}
    toolbarOnFocus={false}
  />
</div>
`;
